core/block/source/source.go

```diff
@@ -27,12 +29,64 @@ import (
 	"github.com/anyproto/anytype-heart/util/slice"
 )
 
-var log = logging.Logger("anytype-mw-source")
+const (
+	defaultDataType  = "1/s"
+	poolSize         = 4096
+	snappyLowerLimit = 64
+)
+
 var (
+	log = logging.Logger("anytype-mw-source")
+
+	bytesPool = sync.Pool{New: func() any { return make([]byte, poolSize) }}
+
 	ErrObjectNotFound = errors.New("object not found")
 	ErrReadOnly       = errors.New("object is read only")
 )
 
+func MarshallChange(c *pb.Change) (res []byte, dataType string, err error) {
+	data := bytesPool.Get().([]byte)[:0]
+	defer bytesPool.Put(data)
+
+	data = slices.Grow(data, c.Size())
+	n, err := c.MarshalTo(data)
+	if err != nil {
+		return
+	}
+	data = data[:n]
+
+	if n > snappyLowerLimit {
+		res = snappy.Encode(nil, data)
+		log.Debugf("change is shrunk by snappy from %d bytes to %d bytes. Space saving: %.2f%%",
+			len(data), len(res), 100*(1-float32(len(res))/float32(len(data))))
+		dataType = defaultDataType
+	} else {
+		res = data
```

Comment:
In the case when you don't use snappy you shouldn't use sync.Pool because these bytes leaks outside this func.
Or you can just copy data to res